training/creating_dataset.py

```diff
@@ -15,117 +13,106 @@
 
 
 def is_path_correct(path: str) -> bool:
     """
     Checks whether given path is correct: file exists and correct file extension - json or db
+
     :param path: path to book to load
-    :return: True if path is correct otherwise False
+    :return: True if path is correct
     """
     if not os.path.isfile(path):
-        print(f"Path {path} does not exist")
-        return False
+        raise ValueError(f"Path {path} does not exist")
+        # return False
     if not (path.endswith(".db") or path.endswith(".json")):
-        print(f"Expected db or json file, got {path}")
-        return False
+        raise ValueError(f"Expected db or json file, got {path}")
+        # return False
     return True
 
 
-def move_values_level_up(value_name: str) -> callable:
+def load_and_preprocess_data(
+    paths_to_books: list[str] = None,
+    characteristics: list[str] = None,
+) -> pd.DataFrame:
     """
-    Assigns values of chosen key in columns' dictionaries to that columns
-    :param value_name: name of value that should be moved one level up as value of current column/columns
-    :return: function which create DataFrame with values of current column(s) changed to value of chosen subcolumn
+    Creates dataframe containing chosen characteristics, level and source book of monsters from chosen books
+
+    :param paths_to_books: list of paths of books to load
+    :param characteristics: list of characteristics to load
+    :return: DataFrame with monsters (NPC) from chosen books and with chosen characteristics and their origin book
     """
 
-    def inner_move_values_level_up(df: pd.DataFrame) -> pd.DataFrame:
-        return df.applymap(lambda x: x.get(value_name), na_action="ignore")
+    if paths_to_books is None:
+        paths_to_books = DATASET_PATHS
 
-    return inner_move_values_level_up
+    if characteristics is None:
+        characteristics = ["abilities", "ac", "hp"]
 
+    data = []
 
-def get_subcolumn(book: pd.DataFrame, subcolumn_path: str) -> pd.DataFrame:
-    """
-    Gets subcolumn of given DataFrame according to given path
-    :param book: DataFrame with all data from book(s)
-    :param subcolumn_path: path to subcolumn
-    :return: chosen subcolumn as DataFrame
-    """
-    subcol = deepcopy(book)
-    for col in subcolumn_path.split("/"):
-        subcol = pd.DataFrame(data=unpack_column(subcol, col))
+    for path in paths_to_books:
+        is_path_correct(path)
+        with open(path) as file:
+            # loading json strings from files
+            d = [json.loads(line) for line in file]
+            # getting only system column (all characteristics are there)
+            # getting only npc - monsters
+            data += [line.get("system") for line in d if line.get("type") == "npc"]
+            file.close()
 
-    return subcol
+    bestiary = pd.json_normalize(data)
 
+    columns = []
+    for characteristic in set(characteristics + ["book", "level"]):
```

Comment:
Zostawiłbym tutaj listę, po co zmieniać na set? I tak po tym iterujemy

Reply:
to na wypadek gdyby użytkownik np. dwa razy wpisał te samą cechę lub book / level i kilka razy wyodrębnimy tę samą kolumnę. Po zastosowaniu `concat` mamy kilka kolumn z tymi samymi wartościami i nazwą. Wydawało mi się, że może nie ma po co zajmować się więcej niż raz jedną cechą